### PR TITLE
MAINT: SciPy 1.6.1 backports

### PR DESCRIPTION
**Pull request by tylerjereddy**
* every merged PR with a backport label is included in this PR
* there are going to be issues that are not yet resolved because of the large number of changes done/needed/pending related to the release of NumPy `1.20.0`
* gh-13518 also needs to be merged in before `1.6.1`, and the resolution of the HiGHS namespacing issue would be nice as well

### CODE REVIEW

**Comment by mckib2** [comment]
If I remember correctly, gh-13354 should be good to go, just waiting on a reviewer to verify the fix and okay

**Comment by tylerjereddy** [comment]
I pulled in the extra backport mentioned above and rebased/force pushed just to see where we are with things, and now that NumPy `1.20.1` is out.

I'm still organizing the release notes/labels a bit, and I imagine a few more backports might be needed to get CI green but we'll see.

**Comment by tylerjereddy** [comment]
@mdhaber @rlucas7 I'm seeing a small number of stats-related test failures as shown below. Looks like they may be from gh-13067; perhaps they snuck in a backport merge conflict somehow and it would be safe for me to purge them out?

```
2021-02-11T04:03:22.4067463Z ================================== FAILURES ===================================
2021-02-11T04:03:22.4068996Z ______________________ TestGeoMean.test_weights_1d_list _______________________
2021-02-11T04:03:22.4070292Z [gw1] win32 -- Python 3.8.7 C:\hostedtoolcache\windows\Python\3.8.7\x64\python.exe
2021-02-11T04:03:22.4071392Z C:\hostedtoolcache\windows\Python\3.8.7\x64\lib\site-packages\scipy\stats\tests\test_stats.py:4617: in test_weights_1d_list
2021-02-11T04:03:22.4072654Z     check_equal_gmean(a, desired, weights=weights, rtol=1e-5)
2021-02-11T04:03:22.4073645Z E   TypeError: check_equal_gmean() got an unexpected keyword argument 'weights'
2021-02-11T04:03:22.4074530Z         a          = [1, 2, 3, 4, 5]
2021-02-11T04:03:22.4075378Z         desired    = 2.77748
2021-02-11T04:03:22.4076317Z         self       = <scipy.stats.tests.test_stats.TestGeoMean object at 0x000001FE2A00F250>
2021-02-11T04:03:22.4077257Z         weights    = [2, 5, 6, 4, 3]
2021-02-11T04:03:22.5066546Z ______________________ TestGeoMean.test_weights_1d_array ______________________
2021-02-11T04:03:22.5068650Z [gw1] win32 -- Python 3.8.7 C:\hostedtoolcache\windows\Python\3.8.7\x64\python.exe
2021-02-11T04:03:22.5070458Z C:\hostedtoolcache\windows\Python\3.8.7\x64\lib\site-packages\scipy\stats\tests\test_stats.py:4625: in test_weights_1d_array
2021-02-11T04:03:22.5071276Z     check_equal_gmean(a, desired, weights=weights, rtol=1e-5)
2021-02-11T04:03:22.5072422Z E   TypeError: check_equal_gmean() got an unexpected keyword argument 'weights'
2021-02-11T04:03:22.5073069Z         a          = array([1, 2, 3, 4, 5])
2021-02-11T04:03:22.5073993Z         desired    = 2.77748
2021-02-11T04:03:22.5074564Z         self       = <scipy.stats.tests.test_stats.TestGeoMean object at 0x000001FE0D516820>
2021-02-11T04:03:22.5075611Z         weights    = array([2, 5, 6, 4, 3])
2021-02-11T04:03:22.5985421Z __________________ TestGeoMean.test_weights_masked_1d_array ___________________
2021-02-11T04:03:22.5987141Z [gw1] win32 -- Python 3.8.7 C:\hostedtoolcache\windows\Python\3.8.7\x64\python.exe
2021-02-11T04:03:22.5989506Z C:\hostedtoolcache\windows\Python\3.8.7\x64\lib\site-packages\scipy\stats\tests\test_stats.py:4633: in test_weights_masked_1d_array
2021-02-11T04:03:22.5990946Z     check_equal_gmean(a, desired, weights=weights, rtol=1e-5)
2021-02-11T04:03:22.5992744Z E   TypeError: check_equal_gmean() got an unexpected keyword argument 'weights'
2021-02-11T04:03:22.5993802Z         a          = array([1, 2, 3, 4, 5, 6])
2021-02-11T04:03:22.5995244Z         desired    = 2.77748
2021-02-11T04:03:22.5996273Z         self       = <scipy.stats.tests.test_stats.TestGeoMean object at 0x000001FE0D50D9D0>
2021-02-11T04:03:22.5997985Z         weights    = masked_array(data=[2, 5, 6, 4, 3, --],
2021-02-11T04:03:22.5998777Z              mask=[False, False, False, False, False,  True],
2021-02-11T04:03:22.5999841Z        fill_value=999999)
```

**Comment by mdhaber** [comment]
Yes, `weights` wasn't a parameter before that PR, so please remove the tests if the keyword isn't there.

**Comment by tylerjereddy** [comment]
Thanks, I've back out those test additions.

I also found time to re-milestone the backport issues/PRs and updating the release notes here accordingly.

**Comment by rlucas7** [comment]
> @mdhaber @rlucas7 I'm seeing a small number of stats-related test failures as shown below. Looks like they may be from [gh-13067](https://github.com/scipy/scipy/pull/13067); perhaps they snuck in a backport merge conflict somehow and it would be safe for me to purge them out?

Yeah I agree with @mdhaber should be safe to purge the changes that added the `weights` keyword to `gmean` for the backport.

**Comment by rlucas7** [comment]
> @mdhaber @rlucas7 I'm seeing a small number of stats-related test failures as shown below. Looks like they may be from [gh-13067](https://github.com/scipy/scipy/pull/13067); perhaps they snuck in a backport merge conflict somehow and it would be safe for me to purge them out?

Yeah I agree with @mdhaber should be safe to purge the changes that added the `weights` keyword to `gmean` for the backport.

**Comment by tylerjereddy** [comment]
@rgommers Ok, I've backported gh-13520; it required some merge conflict resolution and I've assumed that I can mostly just use the new stuff in that config file apart from retaining the old `setuptools` pin on the maintenance branch. There was also some difference about Python 3.9 vs. 3.10, but since both of those Python versions are explicited listed in the new changes I've just used those directly here too.

File: pyproject.toml [hunk 1]
```diff
@@ -3,10 +3,33 @@ requires = [
     "wheel",
     "setuptools<=51.0.0",
     "Cython>=0.29.18",
-    "numpy==1.16.5; python_version=='3.7'",
-    "numpy==1.17.3; python_version=='3.8'",
-    "numpy==1.19.3; python_version=='3.9'",
-    # do not pin numpy on future versions of python to avoid incompatible numpy and python versions
-    "numpy; python_version>='3.10'",
     "pybind11>=2.4.3",
+    "pythran",
```
Comment (tylerjereddy) [inline pyproject.toml:7]: @rgommers hmm, I don't think this belongs in the backport?

Comment (tylerjereddy) [reply pyproject.toml]: I've pulled that out now at least, hopefully nothing else subtle in there that shouldn't be

Comment (rgommers) [reply pyproject.toml]: indeed, that's correct. sorry forgot to mention; we just need the more elaborate numpy version handling

**Comment by tylerjereddy** [comment]
CI is all green, so in it goes

**Pull request merged**